frontend/internal-packages/agent/src/langchain/agents/pmAnalysisAgent/agent.ts

```diff
@@ -1,107 +1,371 @@
+    return Array.isArray(messages) && messages.every(isSearchResultMessage)
+  }
+
+  return true // Allow empty objects or objects without messages
+}
+
+function toSearchResult(value: unknown): SearchResult {
+  if (isSearchResult(value)) {
+    return value
+  }
+  return { messages: [] }
+}
+
+function isToolCall(value: unknown): value is ToolCall {
+  return (
+    isRecord(value) &&
+    'name' in value &&
+    typeof value['name'] === 'string' &&
+    'args' in value &&
+    isRecord(value['args']) &&
+    (value['id'] === undefined || typeof value['id'] === 'string')
+  )
+}
+
+function isModelResponse(value: unknown): value is ModelResponse {
+  if (!isRecord(value)) return false
+
+  // Check tool_calls if present
+  if ('tool_calls' in value) {
+    const toolCalls = value['tool_calls']
+    if (toolCalls !== undefined && !Array.isArray(toolCalls)) return false
+    if (Array.isArray(toolCalls) && !toolCalls.every(isToolCall)) return false
+  }
+
+  // Check additional_kwargs if present
+  if ('additional_kwargs' in value) {
+    const additionalKwargs = value['additional_kwargs']
+    if (additionalKwargs !== undefined && !isRecord(additionalKwargs))
+      return false
+  }
+
+  return true // Allow objects with content property
+}
+
+function toModelResponse(value: unknown): ModelResponse {
+  if (isModelResponse(value)) {
+    const result: ModelResponse = {
+      content: value['content'],
+    }
+
+    if (Array.isArray(value['tool_calls'])) {
+      result.tool_calls = value['tool_calls']
     }
-  >
+
+    if (isRecord(value['additional_kwargs'])) {
+      result.additional_kwargs = value['additional_kwargs']
+    }
+
+    return result
+  }
+  return { content: '', tool_calls: [], additional_kwargs: {} }
+}
+
+export class PMAnalysisAgent {
+  private webSearchTool: DynamicStructuredTool
+  private jsonParser: JsonParser
 
   constructor() {
-    const baseModel = new ChatOpenAI({
-      model: 'o4-mini',
-      reasoning: { effort: 'high', summary: 'detailed' },
-      useResponsesApi: true,
-    })
+    // Store tool separately instead of binding it
+    this.webSearchTool = this.createWebSearchTool()
+    this.jsonParser = new JsonParser()
+  }
+
+  private createWebSearchTool() {
+    return tool(
+      async (input) => {
+        // Type guard for input validation
+        const inputObj: Record<string, unknown> = isRecord(input) ? input : {}
+        const query = String(inputObj['query'])
+
+        const searchResult = await ResultAsync.fromPromise(
+          (async () => {
+            // Use gpt-4o-mini specifically for web search
+            const searchModel = new ChatOpenAI({
+              model: 'gpt-4o-mini',
+              temperature: 0.3,
+            })
+
+            // Create a temporary agent for web search
+            const searchAgent = createReactAgent({
+              llm: searchModel,
+              tools: [{ type: 'web_search_preview' }],
+            })
```

Comment:
I would like to see this properly defined as it violates: https://github.com/humanlayer/12-factor-agents/blob/main/content/factor-02-own-your-prompts.md.

Reply:
I have modified it to not use `createReactAgent`.🙏
https://github.com/liam-hq/liam/pull/2868/commits/e538e28cf9bc812ad96f37a51826ddc347eb8708